Quasardb/Table/Writer/QdbTableExpWriter.cs

```diff
@@ -0,0 +1,383 @@
+        }
+
+        qdb_exp_batch_push_table_data convert_data(qdb_ts_column_info_ex[] infos, qdb_exp_batch_push_column_data[] data)
+        {
+            qdb_exp_batch_push_table_data d;
+
+            d.row_count = (qdb_size_t)_timestamps.Length;
+            d.column_count = (qdb_size_t)infos.Length;
+            d.timestamps = (qdb_timespec*)convert_array(_timestamps);
+            d.columns = (qdb_exp_batch_push_column*)convert_array(convert_columns(infos, data, ref d.column_count));
+            return d;
+        }
+
+        qdb_exp_batch_push_table convert_table(string name, qdb_timespec[] timestamps, qdb_ts_column_info_ex[] infos, qdb_exp_batch_push_column_data[] data)
+        {
+            qdb_exp_batch_push_table table;
+            table.name = convert_string(name);
+            table.data = convert_data(infos, data);
+
+            if (_options.Mode() == qdb_exp_batch_push_mode.truncate)
+            {
+                qdb_ts_range[] ranges = new qdb_ts_range[1];
+                ranges[0] = _options.Interval().ToNative();
+                table.truncate_ranges = (qdb_ts_range*)convert_array(ranges);
+                table.truncate_range_count = (qdb_size_t)1;
+            }
+            else
+            {
+                table.truncate_ranges = null;
+                table.truncate_range_count = (qdb_size_t)0;
+            }
+            table.options = _options.Option();
+            return table;
+        }
+
+        internal QdbTableExpWriter(qdb_handle handle, string table, QdbTableExpWriterOptions options) : base(IntPtr.Zero, true)
+        {
+            _handle = handle;
+            _pins = new List<GCHandle>(1024);
+            _table = table;
+            _options = options;
+
+            using (var columns = new qdb_buffer<qdb_ts_column_info_ex>(handle))
+            {
+                var err = qdb_api.qdb_ts_list_columns_ex(handle, _table, out columns.Pointer, out columns.Size);
+                QdbExceptionThrower.ThrowIfNeeded(err, alias: table);
+
+                long index = 0;
+                _columns = new qdb_ts_column_info_ex[(int)columns.Size];
+                _data = new qdb_exp_batch_push_column_data[(int)columns.Size];
+                foreach (var column in columns)
+                {
+                    _columns[index] = column;
+                    _data[index].blobs = null;
+                    index++;
+                }
+            }
+        }
+
+        /// <inheritdoc />
+        [ReliabilityContract(Consistency.WillNotCorruptState, Cer.Success)]
+        protected override bool ReleaseHandle()
+        {
+            foreach (var pin in _pins)
+                pin.Free();
+            return true;
+        }
+
+        /// <inheritdoc />
+        public override bool IsInvalid
+        {
+            get { return _handle == null || _handle.IsInvalid; }
+        }
+
+        internal long IndexOf(string column)
+        {
+            for (int i = 0; i < _columns.Length; ++i)
```

Comment:
Is this the index of a column within a table, or within all tables of the batch writer? I think the former, right?

Have you considered indexing this structure? It's potentially in a very, very hot code path.

Reply:
done.